### ISSUE TITLE
feat(scripts): description-graph validator + suite self-audit (Phase 3)

### ISSUE THREAD

**Issue by Nosmoht** [Nosmoht/review-claude-config#217]
## Context

Phase 1 (#215) and Phase 2 (#216) establish the empirical foundation and the normative rubric. Phase 3 turns those into **mechanical enforcement**:

1. A catalog-graph validator that detects defects no per-primitive review can catch — name collisions, reciprocal counter-case asymmetries, semantic-cluster competition, aggregate token-budget violations
2. A self-audit run that exercises the validator against the suite's own primitives, surfacing existing description debt as discoverable issues

Without Phase 3, the rubric and convention from Phase 2 are advisory text. With Phase 3, they are CI-enforceable.

**Prior art adopted:**
- mcp-tef — embedding similarity threshold **0.85** for tool overlap detection
- mcp-checkup — token-budget grading scale (per-tool A ≤100 / B ≤300 / C ≤600 / D ≤1000 / F >1000; per-server A ≤500 / B ≤1500 / C ≤3000 / D ≤6000 / F >6000)

## Goal

(a) Implement `scripts/validate_description_graph.py` covering name-collision, reciprocal-symmetry, semantic-cluster, and token-budget checks; wire it into `make validate`. (b) Run the validator against the suite itself and file each finding as a discoverable issue (NO silent fixes).

### A5 — Validator Script

- **A5.1** File exists at `scripts/validate_description_graph.py`
- **A5.2** Script discovers all primitives via filesystem walk: `skills/**/SKILL.md`, `agents/*.md`, `.claude/skills/**/SKILL.md`, `.claude/agents/*.md`, `.claude-plugin/plugin.json`
- **A5.3** **Name-collision check**: detects when two primitives expose the same `name` field (frontmatter or plugin.json); exit code non-zero on collision; report includes both file paths
- **A5.4** **Reciprocal-symmetry check**: parses each description for "Do NOT use; use /X" and "Use /Y instead" patterns; verifies that target primitive contains a matching back-reference; reports asymmetric pairs
- **A5.5** **Semantic-cluster check**: computes pairwise cosine similarity of descriptions using a published embedding model (sentence-transformers/all-MiniLM-L6-v2 or equivalent — declared in script header); pairs with similarity ≥ 0.85 reported as cluster-conflict candidates (per mcp-tef precedent)
- **A5.6** **Per-primitive token-grade**: each description scored A/B/C/D/F per mcp-checkup scale (A ≤100, B ≤300, C ≤600, D ≤1000, F >1000 tokens); F-grade primitives flagged
- **A5.7** **Catalog-aggregate token check**: total description tokens summed; report aggregate against published catalog-degradation thresholds (LongFuncEval 65k–120k danger zone)
- **A5.8** Script supports `--format text` (default) and `--format json` for downstream tooling
- **A5.9** Exit code semantics: 0 = no findings, 1 = warnings only (asymmetry, cluster-conflict, F-grade), 2 = errors (name-collision, aggregate-budget violation)
- **A5.10** Tests exist at `tests/test_validate_description_graph.py` with ≥4 cases: (a) no findings, (b) name collision, (c) asymmetric counter-case, (d) cluster-conflict candidate
- **A5.11** Make target `make validate-descriptions` exists and invokes the script
- **A5.12** `make validate` includes `validate-descriptions` as a step
- **A5.13** Embedding model dependency added to `pyproject.toml` `[project.optional-dependencies].dev` group with pinned version (Reproducibility)

### A6 — Suite Self-Audit

- **A6.1** Validator run against own repo: `python3 scripts/validate_description_graph.py --repo .` produces a report
- **A6.2** Report stored at `docs/audits/description-graph-2026-MM-DD.md` (today's date) with timestamped frontmatter
- **A6.3** Each finding (warning or error) filed as a separate GitHub issue under `category: description-quality` — ONE issue per finding, NOT a single mega-issue
- **A6.4** Issues link back to this Phase 3 issue and include reproducer command
- **A6.5** Repo's own primitives are NOT silently fixed during this phase — fixes are scheduled as separate work
- **A6.6** Calibration check: at least 30% of own primitives flagged with ≥1 warning. If <30%, the rubric is too permissive and a calibration-issue is filed against Phase 2's rubric.

### Cross-Cutting

- **AC.1** `make validate` exits 0 (validator can produce findings without failing CI when warnings only — exit 1 must be tolerated by `make validate` OR findings fixed before merge)
- **AC.2** Conventional commit format
- **AC.3** No hardcoded user-home paths; use `Path.home()` in Python
- **AC.4** No tracker IDs

### In Scope
- Single Python script `validate_description_graph.py` (≥100 LOC justified by Right-Altitude rule — narrowly algorithmic + deterministically testable + parser/AST work)
- Test fixtures + ≥4 test cases
- Make-target wiring
- Self-audit run + report file + filed sub-issues

### Out of Scope (later phases)
- Per-primitive review skill extensions (`/review-skill`, `/review-agent` etc.) → Phase 4
- New `/audit-description-graph` skill (LLM-callable wrapper around the validator) → Phase 4
- Apply-skill extensions for description-smell-fix actions → Phase 4
- `/audit-undertriggering` skill → Phase 5
- Eval cases for the new audit logic → Phase 5

### Never
- Silent self-fixes of own primitives during the audit run. Every finding becomes a tracked issue. Reason: closing the audit loop without filing makes the audit invisible to maintainers and consumers.
- Embedding model unpinned. Reason: cluster-similarity is reproducibility-sensitive; cosine ≥ 0.85 only matters if the embedding space is fixed.

## Dependencies & References

**Blocks:** Phase 4 (#TBD), Phase 5 (#TBD)
**Blocked by:** Phase 1 (#215), Phase 2 (#216)

**Prior art:**
- [mcp-tef — testing tool descriptions before deployment](https://dev.to/stacklok/introducing-mcp-tef-testing-your-mcp-tool-descriptions-before-they-cause-problems-fan)
- [mcp-checkup — token-cost grading](https://github.com/yifanyifan897645/mcp-checkup)

**Repo references:**
- `scripts/validate_token_budgets.py` (existing pattern for validators)
- `scripts/validate_schema.py` (existing pattern)
- `Makefile` (existing `validate` target)
- `tests/` (existing pytest layout)
- `pyproject.toml` `[project.optional-dependencies].dev` (where embedding dep goes)

## Verification

```bash
# A5
test -f scripts/validate_description_graph.py
python3 scripts/validate_description_graph.py --help
python3 scripts/validate_description_graph.py --format json | jq -e '.findings | type == "array"'
test -f tests/test_validate_description_graph.py
pytest tests/test_validate_description_graph.py -v
make validate-descriptions
make validate

# A6
ls docs/audits/description-graph-*.md | head -1
gh issue list --repo Nosmoht/review-claude-config --label "category: description-quality" --search "in:body 'reproducer'" | wc -l  # ≥1 (filed sub-issue)
```

## Risk

**Class:** medium — adds CI dependency (embedding model) and surfaces existing debt that must then be triaged.

**R5 sub-heuristic check:**
- *Risk-mitigation:* per-primitive reviews cannot detect catalog-level defects — name collisions, reciprocal asymmetries, and cluster competition all require global view. Defect class: "catalog-graph silent rot."
- *Existing-coverage subtraction:* `validate_token_budgets.py` covers token budgets per file but NOT aggregate-catalog or graph properties. `validate_schema.py` covers structural schema, not semantic graph. Non-redundant.
- *Test-the-test inversion:* test fixtures must include both passing and FAILING cases (A5.10) to verify the validator detects what it claims. Pure pass-cases would let the validator silently ignore problems.
- *Structure-insensitive:* AC checks invoke the validator and assert exit codes / output shape, not byte-equal output snapshots. Embedding-similarity check is semantic, not lexical.

**Reproducibility risk:** embedding-similarity threshold 0.85 was set by mcp-tef on an unspecified model. We pin the model to make our own threshold reproducible (A5.13). Threshold may need calibration once we run on the suite — calibration-issue is the planned fallback (A6.6).

---

🤖 Phase 3 of the description-quality track. Builds on Phase 1 (#215) reference + Phase 2 (#216) rubric.

**Comment by Nosmoht** [comment]
**FOSS-only constraint (no commercial build dependencies):**

Per repo policy (memory: keine kommerziellen Build-Dependencies, post-#199/#213), the embedding model in **A5.5 / A5.13** must be FOSS and runnable locally — no SaaS embedding APIs (Voyage / OpenAI Embeddings / Cohere / Jina). Acceptable choices:

- `sentence-transformers/all-MiniLM-L6-v2` (Apache-2.0, ~22M params, local inference, model weights from HuggingFace Hub at install time — no runtime API calls)
- Stdlib-only fallback: TF-IDF cosine via `sklearn.feature_extraction.text.TfidfVectorizer` + `sklearn.metrics.pairwise.cosine_similarity` (already-FOSS)
- Pure-stdlib fallback: token-overlap Jaccard similarity (no extra dep)

Builder picks one at plan time and pins exact version in `pyproject.toml`. Threshold 0.85 (mcp-tef precedent) may need re-calibration depending on the chosen embedding space — calibration falls under A6.6's calibration-issue fallback.

**Comment by Nosmoht** [comment]
**Description-Quality Track context** (deep-research session 2026-05-07):

- Phase 1 — Foundation (A1 + A4): #215
- Phase 2 — Standards (A2 + A3): #216
- Phase 3 — Validation (A5 + A6): #217
- Phase 4 — Audit Extensions (B1 + B2 + B4): #218
- Phase 5 — Closeout (B3 + B5): #219

Dependency chain: #215 → #216 → #217 → #218 → #219.
Track label: `category: description-quality`.
Issue bodies were drafted with phase-numbers as `#TBD` placeholders; this comment supersedes those placeholders.

**Comment by Nosmoht** [comment]
**Description-Quality Track context** (revised 2026-05-07 after #223 closure):

- Phase 1 — Foundation Reference (A1, includes empirical anchors): #215
- Phase 2 — Standards (A2 + A3): #216
- Phase 3 — Validation (A5 + A6): #217
- Phase 4 — Audit Extensions (B1 + B2 + B4): #218
- Phase 5 — Closeout (B3 + B5): #219

**#223 closed as superseded** (architectural mismatch — see #223 closure rationale). Empirical anchors (97.1% smell rate, +5.85pp augmentation, 7–85% catalog degradation) live in `description-design-problem.md` (#215, A1.7 + A1.7b) as `[Proven result]` citations, not in `engineering-baseline.md`. The baseline is general-purpose engineering technique reference; description-quality is domain-specific knowledge.

**Evidence-class correction:** All empirical anchors carry canonical class label `[Proven result]` per `references/evidence-contract.md`. The earlier issue bodies used non-canonical `[Validated — peer-reviewed]` — that label has no contract status and must not be used.

Dependency chain: #215 → #216 → #217 → #218 → #219.
Track label: `category: description-quality`.

**Comment by Nosmoht** [comment]
## Pipeline evaluation results

**Verdict**: PASS
**Phase**: 7.5 (builder-evaluator)
**Commit**: 2d48d12 (HEAD of `feat/issue-217-description-graph-validator`; PR #250)
**Builder commit**: bcc8254
**CI**: 5/5 GREEN (Ruff Lint+Format / Policy Consistency / Schema Validation / Tests / Description-Graph Validator)

### Acceptance criteria
- Total: 23 | Passed: 23 | Failed: 0
- A5.1–A5.13 (validator script): all PASS
- A6.1–A6.6 (suite self-audit): all PASS — 22 sub-issues filed (#228–#249), flag rate 68.2% ≥ 30% calibration threshold
- AC.1–AC.4 (cross-cutting): all PASS

### Findings
- **CRITICAL (0)**: none
- **WARNING (3)**:
  1. A5.13 issue AC text says "dev group"; plan overrides to separate `[project.optional-dependencies].description-graph` extra. Architecturally sounder; predicate passes via plan-revised regex.
  2. A5.2 uses `skills/*/SKILL.md` (single-level) — no files missed at current repo depth; predicate passes (44 ≥ 40).
  3. `EMBEDDING_REVISION` is documentation anchor only — `model2vec` upstream does not support `revision=` kwarg pinning. Limitation explicitly documented at `scripts/validate_description_graph.py:158–159`.
- **INFO (1)**:
  1. `main()` exit code 0 not exercised end-to-end in CI (repo always has cluster warnings on real catalog); unit-tested at check-function level in test #1.

### Hard-Constraint sweep
- HC1 (no hardcoded home-dir prefixes): PASS
- HC2 (no external tracker IDs): PASS
- HC3 (`make validate` exits 0): PASS — 1088 tests pass, 6 pre-existing DeprecationWarnings
- HC4 (Conventional-commit format): PASS — both commits match
- HC5 (no `--no-verify` / `--no-gpg-sign`): PASS
- HC6 (no mid-session edits to scoring-rubric / engineering-baseline): PASS — neither file in diff
- HC7 (no internal IPs/hostnames in public artifacts): PASS — `scrub()` filter + manual review

### Disposition
PASS — proceed to close via PR #250 merge

Findings file: `.work/issue-217/evaluator-findings.md` (gitignored locally; full detail in this comment for audit trail)
